cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -242,6 +245,47 @@ private static Job getNamedJob(JobContext context, String namedOutput) throws IO
     return job;
   }
 
+  /**
+   * Wraps RecordWriter to increment output counters.
+   */
+  @SuppressWarnings("unchecked")
+  private static class RecordWriterWithCounter extends RecordWriter {
+    private final RecordWriter writer;
+    private final String groupName;
+    private final String counterName;
+    private final TaskInputOutputContext context;
+
+    public RecordWriterWithCounter(RecordWriter writer, TaskInputOutputContext context) {
+      this.writer = writer;
+      this.context = context;
+      this.groupName = TaskCounter.class.getName();
+      this.counterName = getCounterName(context);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void write(Object key, Object value) throws IOException, InterruptedException {
+      context.getCounter(groupName, counterName).increment(1);
+      writer.write(key, value);
+    }
+
+    public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+      writer.close(context);
+    }
+
+    private String getCounterName(TaskInputOutputContext context) {
+      MapReduceMetrics.TaskType taskType = MapReduceMetrics.TaskType.from(context.getTaskAttemptID().getTaskType());
+      Preconditions.checkNotNull(taskType, "Illegal task type: " + context.getTaskAttemptID().getTaskType());
```

Comment:
It's better for the `TaskType.from` to throw `IllegalArgumentException` instead of requiring caller to check for `null`. Remember, `null` is (kind of) evil.